core/trino-main/src/main/java/io/trino/operator/ExchangeOperator.java

```diff
@@ -197,6 +215,290 @@ public Page getOutput()
+    {
+        private final DirectExchangeClient directExchangeClient;
+
+        private DirectExchangeDataSource(DirectExchangeClient directExchangeClient)
+        {
+            this.directExchangeClient = requireNonNull(directExchangeClient, "directExchangeClient is null");
+        }
+
+        @Override
+        public SerializedPage pollPage()
+        {
+            return directExchangeClient.pollPage();
+        }
+
+        @Override
+        public boolean isFinished()
+        {
+            return directExchangeClient.isFinished();
+        }
+
+        @Override
+        public ListenableFuture<Void> isBlocked()
+        {
+            return directExchangeClient.isBlocked();
+        }
+
+        @Override
+        public void addSplit(RemoteSplit remoteSplit)
+        {
+            DirectExchangeInput taskInput = (DirectExchangeInput) remoteSplit.getExchangeInput();
+            directExchangeClient.addLocation(taskInput.getTaskId(), taskInput.getLocation());
+        }
+
+        @Override
+        public void noMoreSplits()
+        {
+            directExchangeClient.noMoreLocations();
+        }
+
+        @Override
+        public OperatorInfo getInfo()
+        {
+            return directExchangeClient.getStatus();
+        }
+
+        @Override
+        public void close()
+        {
+            directExchangeClient.close();
+        }
+    }
+
+    private static class ExternalExchangeDataSource
+            implements ExchangeDataSource
+    {
+        private final ExchangeSource exchangeSource;
+        private final LocalMemoryContext systemMemoryContext;
+
+        private ExternalExchangeDataSource(ExchangeSource exchangeSource, LocalMemoryContext systemMemoryContext)
+        {
+            this.exchangeSource = requireNonNull(exchangeSource, "exchangeSource is null");
+            this.systemMemoryContext = requireNonNull(systemMemoryContext, "systemMemoryContext is null");
+        }
+
+        @Override
+        public SerializedPage pollPage()
+        {
+            Slice data = exchangeSource.read();
+            systemMemoryContext.setBytes(exchangeSource.getSystemMemoryUsage());
+            if (data == null) {
+                return null;
+            }
+            // TODO: Avoid extra memory copy
+            return readSerializedPage(data.getInput());
+        }
+
+        @Override
+        public boolean isFinished()
+        {
+            return exchangeSource.isFinished();
+        }
+
+        @Override
+        public ListenableFuture<Void> isBlocked()
+        {
+            return asVoid(toListenableFuture(exchangeSource.isBlocked()));
+        }
+
+        @Override
+        public void addSplit(RemoteSplit remoteSplit)
+        {
+            // ignore
+        }
+
+        @Override
+        public void noMoreSplits()
+        {
+            // ignore
+        }
+
+        @Override
+        public OperatorInfo getInfo()
+        {
+            return null;
```

Comment:
why null?

Reply:
I didn't have a chance to think more about what statistics do we want to expose for external exchange. It feels like it would be more clear once we have some (even minimal) operational experience.

Reply:
There is an item related to that
![image](https://user-images.githubusercontent.com/5570988/147164680-267ac8cc-e3ea-4df8-bb1b-3671645b1df6.png)